lib/expand/phases/etcd.go

```diff
@@ -0,0 +1,285 @@
+		endpoints = append(endpoints, fmt.Sprintf("https://%v:%v",
+			master.AdvertiseIP, defaults.EtcdAPIPort))
+	}
+	stateDir, err := state.GetStateDir()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	etcdClient, err := clients.EtcdMembers(&clients.EtcdConfig{
+		Endpoints:  endpoints,
+		SecretsDir: state.SecretDir(stateDir),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	logger := &fsm.Logger{
+		FieldLogger: logrus.WithFields(logrus.Fields{
+			constants.FieldPhase: p.Phase.ID,
+		}),
+		Key:      opKey(p.Plan),
+		Operator: operator,
+		Server:   p.Phase.Data.Server,
+	}
+	return &etcdExecutor{
+		FieldLogger:    logger,
+		Etcd:           etcdClient,
+		Runner:         runner,
+		Master:         *p.Phase.Data.Master,
+		ExecutorParams: p,
+	}, nil
+}
+
+type etcdExecutor struct {
+	// FieldLogger is used for logging
+	logrus.FieldLogger
+	// Etcd is client to the cluster's etcd members API
+	Etcd etcd.MembersAPI
+	// Runner is used to run remote commands
+	Runner fsm.AgentRepository
+	// Master is one of the master nodes
+	Master storage.Server
+	// ExecutorParams is common executor params
+	fsm.ExecutorParams
+}
+
+// Execute adds the joining node to the cluster's etcd cluster
+func (p *etcdExecutor) Execute(ctx context.Context) error {
+	p.Progress.NextStep("Adding etcd member")
+	member, err := p.Etcd.Add(ctx, fmt.Sprintf("https://%v:%v",
+		p.Phase.Data.Server.AdvertiseIP, defaults.EtcdPeerPort))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	p.Infof("Added etcd member: %v.", member)
+	return nil
+}
+
+// Rollback removes the joined node from the cluster's etcd cluster
+func (p *etcdExecutor) Rollback(ctx context.Context) error {
+	p.Progress.NextStep("Restoring etcd data")
+	backupPath := getBackupPath(p.Plan.OperationID)
+	agentClient, err := p.Runner.GetClient(ctx, p.Master.AdvertiseIP)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = p.checkBackup(ctx, agentClient, backupPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = p.stopEtcd(ctx, agentClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = p.wipeEtcd(ctx, agentClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = p.startEtcd(ctx, agentClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = p.restoreEtcd(ctx, agentClient, backupPath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	p.Infof("Restored etcd data from %v.", backupPath)
+	return nil
+}
+
+func (p *etcdExecutor) checkBackup(ctx context.Context, agent rpcclient.Client, backupPath string) error {
+	var out bytes.Buffer
+	err := agent.Command(ctx, p.FieldLogger, &out, utils.PlanetEnterCommand(
+		defaults.StatBin, backupPath)...)
+	if err != nil {
+		return trace.Wrap(err, "failed to check backup file %v: %s", backupPath, out.String())
+	}
+	return nil
+}
+
+func (p *etcdExecutor) stopEtcd(ctx context.Context, agent rpcclient.Client) error {
+	var out bytes.Buffer
+	err := agent.Command(ctx, p.FieldLogger, &out, utils.PlanetEnterCommand(
```

Comment:
`RunPlanetCommand` executes the command locally while I only need to construct its arguments since it's run on a remote node via an agent. And I couldn't use `PlanetCommandSlice` because it uses local gravity binary path.